### ISSUE TITLE
Set Focus (or Accessibility Focus?) method Creates WCAG Violations

### ISSUE THREAD

**Issue by chriscm2006** [appt-nl/accessibility-android#1]
Following is some friendly constructive criticism and an offer to help! :)

Quick intro: Name's Chris McMeeking (Linked In me!). Graduated from University of Michigan with a degree in Computer Science and Engineering in 2010. Been working in the Native Mobile Accessibility field since, including time advising the W3C's Native Mobile Accessibility Task force on how Native Operating Systems related to WCAG 2.1. 

First, this effort is awesome and I love what you're trying to do. That said, there are some significant issues that need addressing. Following is one concrete example, but there are many: 

Moving input focus and moving accessibility focus are different! This method is misleading. Hard to tell which you intended... otherwise this would be a Pull Request!!! :) Here is the code I found that claims to be moving "Accessibility Focus"

```
/**
 * Moves the accessibility focus to the given view
 *
 * @param view View to move accessibility focus to
 */
fun Accessibility.Companion.setFocus(view: View) {
    view.isFocusable = true
    view.isFocusableInTouchMode = true
    view.sendAccessibilityEvent(AccessibilityEvent.TYPE_VIEW_FOCUSED)
}
```
This code is not moving Accessibility focus. In fact it is:

- Forcing a view to be Input Focusable. (Bad, creates Keyboard User problems)
- Calls a useless method in relation to this for Touch Mode.
- Set Input focus to this view.
- YES, Input Focus will drag Accessibility Focus with it.

This approach to moving Accessibility Focus is dangerous for a myriad of WCAG Specific reasons, particularly those related to keyboard and switch control users. We have pretty negatively impacted their experience and modified your library user's view hierarchy in a way that makes the View overall less accessible.

Forcing a view to be **Input Focusable** in order to move **Accessibility Focus** there OR WORSE leveraging **Input Focus** to move **Accessibility Focus** around creates accessibility problems for users. In particular Keyboard Only users who are suddenly moving input focus to purely informative views. Also for Switch Control users who are now highlighting views without understanding why that view needs highlighted.

This is a frequently used hack by engineers who either don't know how to move only Accessibility Focus or who don't understand why a view won't accept Accessibility Focus when they move it properly!

The correct code for the function as documented would be:

```
/**
 * Moves the accessibility focus to the given view
 *
 * @param view View to move accessibility focus to
 */
fun Accessibility.Companion.setFocus(view: View) {
    view.sendAccessibilityEvent(AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUSED)
}
```

I see similar problems scattered throughout this library. Please feel free to reach out on LinkedIn. Happy to help! If you partner with me to fix these problems I'd love to share this effort with my colleagues at CSUN and other places or write a blog post about it! 

Chris

**Comment by JJdeGroot** [comment]
Hey @chriscm2006 ,

Thanks for reaching out with detailed feedback.  It would be great to work together on this library (and other code examples). The library hasn't updated for a while, and the most recent version 0.2 is still a beta version.

With that being said, the goal of the setFocus method is to move the focus of **TalkBack** to the specified `View`. The `isFocusable` and `isFocusableInTouchMode` are needed on some Android versions to actually make this work.

Between the time that this library was published and now, I have also realised that these properties have negative side effects for other assistive technologies, such as switch access and the external keyboard.

I also agree with the points you make about worsening the user experience for these users. This method was purely written with screen reader users in mind. The method name should be changed to reflect this and/or modified with a check which assistive technologies are being used.

It looks like we were already connected on LinkedIn, I've sent you a message.

We recently published a lot of code examples, including for React Native, Flutter and Xamarin. The idea behind this library was to unify the syntax for all of these programming languages and frameworks. We now have the code examples to actually achieve this.

But I/we have also realised that maintenance is an important part of a library. Which is why we are looking for other people to help with maintaining code examples and to think about an accessibility library such as this one.

Looking forward to have a chat in the near future!
